doc/statemachines.tex

```diff
@@ -0,0 +1,310 @@
+  The state machine is an abstract high-level description of the network protocol.
+
+\item[What questions, properties can be reasoned about in this framework ? (the typed transition framework ?)]
+  Client and server agree on the current state (in particular on the Done-state) ?
+  There are no deadlocks.
+
+\item[Model of Concurrency ?]
+  Haskell implementation (typed transition framework):
+  The typed transition framework is based on handshake synchronization (similar to the communicating sequential processes/ process algebra)).
+  An event is a single point in time when the system transitions from one state to an other state.
+  The type system does not represent an intermediate state. It does not model messages that are send at time
+  $t_1$ and arrive later at time $t_2$.
+  
+  Real world systems use message passing. When going from an event-based system (Haskell implementation/ Protocol/Core.hs)
+  
+  to message passing the interpretation/ semantics has to be adapted.
+\item[Timeouts, lost messages, etc]
+  Important aspects are not covered at this level/layer.
+\end{description}
+  
+\subsubsection{Notation} \marginpar{\njd{Question: how formal should it be ?}}
+
+  \marginpar{\njd{This part should be generic for all Protocols that are implemented via the typed-transition framework.}}
+
+  Todo: Definition of the terms used in the document:
+  What are the terms that we use to describe the network layer ?
+  Can we reference a common source / textbook ?
+\begin{description}
+\item[Protocol]
+\item[Node]  
+\item[Agent]
+  A unit of execution that follows particular Protocol.
+  (Could be implemented as a green thread for example.)
+  A Node can run many agents.
+\item[Client] The agent that is active in the initial state.
+\item[Server] The agent that waits for a message clients when it starts.  
+\item[Upstream,Downstream,Topology]
+\end{description}
+
+\begin{description}
+\item[State Machine]
+  The Protocols are described as a state machine.
+  Formally the state machine is a tuple $(S,T,initial,final)...$ (States, Transitions, initial state, finial state)...
+  The specification uses different representations of the state machine that all describe the same protocol
+  but may leave out some aspects.
+  \begin{itemize}
+  \item The state machine can be represented as a collection of tables (States, Transitions/Messages..).
+  There is a one to one correspondence between these tables and the Haskell source.
+\item A transition graph of the global view of the state machine.
+\item A transition graph of the client view.
+\item A transition graph of the server view.
+\end{itemize}
+  
+\item[States]
+  The specification assumes a global view of the state of a system consisting of the client and the server.
+  Client and server are always in identical states (by definition)
+  which also means that client and server simultaneously transit to new states.
+  The global state is not observable by the client nor the server.
```

Comment:
In the framework we are using the client and the server agrees on states, and the client server are always in the dual state (only when a message is in flight, the receiver does not know yet it's state, but the sender knows it already).
If you mean by global state: state of all local mutable variables, then that's true.